File: 138. Copy List with Random Pointer.go
/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Next *Node
 *     Random *Node
 * }
 */

func copyRandomList(head *Node) *Node {
    if head == nil {
        return nil
    }
    nodeMap := make(map[*Node]*Node)
    newHead := &Node{
        Val: head.Val,
    }
    nodeMap[head] = newHead
    pointer := head.Next
    current := newHead
    for pointer != nil {
        current.Next = &Node{
            Val: pointer.Val,
        }
        current = current.Next
        nodeMap[pointer] = current
        pointer = pointer.Next
    }
    pointer = head
    current = newHead
    for pointer != nil {
        random := pointer.Random
        current.Random = nodeMap[random]
        current, pointer = current.Next, pointer.Next
    }
    return newHead
}

File: 138. Copy List with Random Pointer.java
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

class Solution {
    public Node copyRandomList(Node head) {
        Map<Node, Node> mapping = new HashMap<>();
        Node pointer = head;
        while (pointer != null) {
            mapping.put(pointer, new Node(pointer.val));
            pointer = pointer.next;
        }
        pointer = head;
        while (pointer != null) {
            Node node = mapping.get(pointer);
            if (pointer.next != null) {
                node.next = mapping.get(pointer.next);
            }
            if (pointer.random != null) {
                node.random = mapping.get(pointer.random);
            }
            pointer = pointer.next;
        }
        return mapping.get(head);
    }
}